[SCEV] BECount to zero if `((-C + (C smax %x)) /u %x), C > 0` holds

antoniofrighetto opened the pull request:
The SCEV expression `((-C + (C smax %x)) /u %x)` can be folded to zero for any positive constant C.

Proof: https://alive2.llvm.org/ce/z/_dLm8C.

nikic commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -3547,6 +3547,25 @@ const SCEV *ScalarEvolution::getUDivExpr(const SCEV *LHS,
+    for (const SCEV *Op : AE->operands()) {
+      if (Op->isAllOnesValue()) {
+        FoundMinusOne = true;
+      } else if (const auto *MME = dyn_cast<SCEVMinMaxExpr>(Op)) {
```
Directly check SCEVSMinExpr?

nikic commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -3547,6 +3547,25 @@ const SCEV *ScalarEvolution::getUDivExpr(const SCEV *LHS,
+  if (const auto *AE = dyn_cast<SCEVAddExpr>(LHS);
+      AE && AE->getNumOperands() == 2 && AE->hasNoSignedWrap()) {
+    bool FoundMinusOne = false, FoundOneSMaxRHS = false;
+    for (const SCEV *Op : AE->operands()) {
```
The constant operand is guaranteed to be first. You don't need to loop or use hasOperand, you know which operand will be which value.

nikic commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -3547,6 +3547,25 @@ const SCEV *ScalarEvolution::getUDivExpr(const SCEV *LHS,
+    }
+
+    if (FoundMinusOne && FoundOneSMaxRHS)
+      // The backedge is never taken.
```
This code is not backedge specific.

nikic commented on llvm/test/Analysis/ScalarEvolution/udiv-of-x-xsmaxone-fold.ll:
```diff
@@ -0,0 +1,78 @@
+
+@g_var = external local_unnamed_addr global i32, align 4
+
+define void @test(i32 noundef %x) {
```
This test looks way too complicated.

antoniofrighetto replied:
Updated by letting LV perform the actual transformation, think this should make sense.

fhahn replied:
Would be great to avoid having a SCEV analysis test depend on running LV. Is there anything that makes it difficult to write a simple IR test that contains an `(-1 + (1 smax %x))<nsw> /u %x)` expression?

antoniofrighetto replied:
Exceptionally, I think the main point here was to demonstrate how LV itself generates that expression from a relatively simple loop, instead of manually handcrafting ourselves an IR with a bit of an unnatural / artificial expression; ergo, showing where the expression actually comes from. I understand the concern though, I'm happy to change the IR if we don't want to have this.

antoniofrighetto replied:
Test rewritten so as to avoid the dependence with LV.

nikic reviewed (commented):
Can you please add an alive2 proof to the PR description?

nikic reviewed (commented):
Looks like this generalizes to any positive constant: https://alive2.llvm.org/ce/z/aS3SUe

nikic commented:
> Looks like this generalizes to any positive constant: https://alive2.llvm.org/ce/z/aS3SUe

We don't need nsw as a precondition either, it's implied: https://alive2.llvm.org/ce/z/_dLm8C

antoniofrighetto commented:
> We don't need nsw as a precondition either, it's implied: https://alive2.llvm.org/ce/z/_dLm8C

Generalized and proof updated, thanks.

antoniofrighetto commented:
> Looks like this generalizes to any positive constant: https://alive2.llvm.org/ce/z/aS3SUe

FWIW, I was unable to generalize the original snippet to let LV generate this expression for other positive constants, which makes me dubious on whether this generalization is worth. But I guess this should be fine as far as compile-time is acceptable (also not because I couldn’t derive this from C snippets doesn't mean that other cases cannot exist where this might be relevant).

antoniofrighetto commented:
Kind ping.

nikic commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -3547,6 +3547,25 @@ const SCEV *ScalarEvolution::getUDivExpr(const SCEV *LHS,
     }
   }
 
+  // ((-C + (C smax %x)) /u %x) evaluates to zero, for any positive %x and any
+  // positive constant C.
```
```suggestion
  // ((-C + (C smax %x)) /u %x) evaluates to zero, for any
  // positive constant C.
```
Positive %x not required?

nikic commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -3547,6 +3547,25 @@ const SCEV *ScalarEvolution::getUDivExpr(const SCEV *LHS,
+      AE && AE->getNumOperands() == 2) {
+    if (const auto *VC = dyn_cast<SCEVConstant>(AE->getOperand(0))) {
+      APInt NegC = VC->getAPInt();
+      if (NegC.isNegative() &&
+          NegC != APInt::getSignedMinValue(NegC.getBitWidth())) {
```
```suggestion
      if ((-NegC).isStrictlyPositive()) {
```
I might phrase it like this, to align more closely with the comment/proof. But your version is fine as well.

antoniofrighetto replied:
Exceptionally left mine as I deem it as slightly more intuitive.

nikic commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -3547,6 +3547,25 @@ const SCEV *ScalarEvolution::getUDivExpr(const SCEV *LHS,
+  if (const auto *AE = dyn_cast<SCEVAddExpr>(LHS);
+      AE && AE->getNumOperands() == 2) {
+    if (const auto *VC = dyn_cast<SCEVConstant>(AE->getOperand(0))) {
+      APInt NegC = VC->getAPInt();
```
```suggestion
      const APInt &NegC = VC->getAPInt();
```

nikic commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -3547,6 +3547,25 @@ const SCEV *ScalarEvolution::getUDivExpr(const SCEV *LHS,
+        if (MME && MME->getNumOperands() == 2) {
+          if (isa<SCEVConstant>(MME->getOperand(0)) &&
+              cast<SCEVConstant>(MME->getOperand(0))->getAPInt() == -NegC)
+            if (MME->getOperand(1) == RHS)
```
Merge the three ifs.

nikic commented on llvm/test/Analysis/ScalarEvolution/udiv-of-x-xsmaxone-fold.ll:
```diff
@@ -0,0 +1,52 @@
+  %g_var.promoted = load i32, ptr @g_var, align 4
+  %smax = tail call i32 @llvm.smax.i32(i32 %x, i32 1)
+  %add = add nsw i32 %smax, -1
+  %udiv = udiv i32 %add, %x
```
Why all the code after the udiv?

antoniofrighetto replied:
The backedge-count explicitly printed zero when was coupled with LV, but even in that case, not needed.

nikic commented on llvm/test/Analysis/ScalarEvolution/udiv-of-x-xsmaxone-fold.ll:
```diff
@@ -0,0 +1,25 @@
+; CHECK-NEXT:  Determining loop execution counts for: @test
+;
+entry:
+  %g_var.promoted = load i32, ptr @g_var, align 4
```
`%g_var.promoted` and the global var are unused.

nikic commented on llvm/test/Analysis/ScalarEvolution/udiv-of-x-xsmaxone-fold.ll:
Can you please also add a test with non-1 constant, and some negative tests (e.g. mismatch between the constants, and negative constant)?

antoniofrighetto replied:
Added, thanks!

nikic commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -3547,6 +3547,23 @@ const SCEV *ScalarEvolution::getUDivExpr(const SCEV *LHS,
+    if (const auto *VC = dyn_cast<SCEVConstant>(AE->getOperand(0))) {
+      const APInt &NegC = VC->getAPInt();
+      if (NegC.isNegative() &&
+          NegC != APInt::getSignedMinValue(NegC.getBitWidth())) {
```
```suggestion
          !NegC.isMinSignedValue()) {
```
Okay, in that case we should adjust it like this.

nikic commented on llvm/test/Analysis/ScalarEvolution/udiv-of-x-xsmaxone-fold.ll:
```diff
@@ -0,0 +1,96 @@
+}
+
+; Negative test, constants mismatch: (-3 + (2 smax %x)) /u %x
+define i32 @test2(i32 %x) {
```
```suggestion
define i32 @test_constant_mismatch(i32 %x) {
```
etc. try to give more meaningful names.

antoniofrighetto replied:
Definitely!

nikic reviewed (approved):
LGTM

antoniofrighetto committed:
[SCEV] BECount to zero if `((-C + (C smax %x)) /u %x), C > 0` holds

The SCEV expression `((-C + (C smax %x)) /u %x)` can be folded
to zero for any positive constant C.

Proof: https://alive2.llvm.org/ce/z/_dLm8C.
